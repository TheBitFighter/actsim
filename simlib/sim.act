/*************************************************************************
 *
 *  This file is part of ACT standard library
 *
 *  Copyright (c) 2022 Rajit Manohar
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 **************************************************************************
 */
namespace sim {

export template<pint W, V>
defproc source (chan!(int<W>) O)
{
  chp {
    *[ O!V ]
  }
}

export template<pint W; pbool REP; pint N; pint data[N]>
defproc source_seq(chan!(int<W>) O)
{
  int i;
  chp {
    *[ i := 0; *[ i < N -> [([]k:N: i=k -> O!data[k])]; i := i + 1 ] <- REP ]
  }
}
  
export template<pbool LOG; pint W>
defproc sink(chan?(int<W>) I)
{
  int x;
  chp {
    *[ I?x; [LOG -> log("sink: ", x, "%x (0x", x, ")") [] else -> skip ] ]
  }
}

export function file_read(int<32> idx) : int<64>;
export function file_eof(int<32> idx) : bool;
export function file_close(int<32> idx) : bool;

/* this assume that the file has at least one value in it */
export template<pint ID; pbool LOOP; pint W>
defproc file_source(chan!(int<W>) O)
{
  bool dummy;
  chp {
    *[ *[ O!file_read(ID) <- ~file_eof (ID) ]; dummy := file_close(ID) <- LOOP ];
     log ("Source file #", ID, " ends.")
  }
}

export template<pint ID; pbool LOOP; pint W>
defproc check_sink(chan?(int<W>) I)
{
  int<W> x, y;
  bool dummy;
  chp {
    *[ *[ I?x;
          y := file_read(ID);
         [ x = y -> skip
         [] else -> log ("ASSERTION failed, value mismatch; expected: ", y, "%x (0x", y, "); got: ", x, "%x (0x", x, ")")
         ]
	  <- ~file_eof(ID)
       ] <- LOOP
    ];
    log ("Sink file #", ID, " ends.")
  }
}

}
