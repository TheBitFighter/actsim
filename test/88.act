import sim;

defproc test()
{
   int idx, i;
   bool tst;
   chp {
       i := 0, idx := sim::rand::init (16);
       tst := sim::rand::seed (idx, 0);
     *[ i < 10 -> 
	 log (i, "%x got: ", sim::rand::get(idx));
         i := i + 1 
      ]
   }
}
