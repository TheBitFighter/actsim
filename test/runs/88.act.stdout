[                  20] <>  0 got: 7304
[                  30] <>  1 got: aa1f
[                  40] <>  2 got: 775a
[                  50] <>  3 got: 95ac
[                  60] <>  4 got: 7a5f
[                  70] <>  5 got: c279
[                  80] <>  6 got: f410
[                  90] <>  7 got: 67af
[                 100] <>  8 got: 7504
[                 110] <>  9 got: d21d
